File: battle/entity/Sniper.java
package common.battle.entity;

import common.CommonStatic;
import common.battle.BattleField;
import common.battle.StageBasis;
import common.battle.attack.AtkModelAb;
import common.battle.attack.AttackAb;
import common.battle.attack.AttackSimple;
import common.pack.UserProfile;
import common.system.P;
import common.system.fake.FakeGraphics;
import common.util.anim.EAnimD;
import common.util.pack.EffAnim.SniperEff;
import common.util.unit.Trait;

import java.util.ArrayList;

public class Sniper extends AtkModelAb {

	private final EAnimD<?> anim = effas().A_SNIPER.getEAnim(SniperEff.IDLE);
	private final EAnimD<?> atka = effas().A_SNIPER.getEAnim(SniperEff.ATK);
	private int coolTime = SNIPER_CD, preTime = 0, atkTime = 0;
	private Entity target;
	public boolean enabled = true, canDo = true;
	public double pos, layer, height, bulletX, targetAngle = 0, cannonAngle = 0, bulletAngle = 0;
	public final BattleField bf; //Used for replay pos/siz gathering

	public Sniper(StageBasis sb, BattleField bf) {
		super(sb);
		this.bf = bf;
	}

	/**
	 * base part of animation
	 */
	public void drawBase(FakeGraphics gra, P ori, double siz) {
		height = ori.y;

		anim.ent[1].alter(9, 1500);
		anim.ent[1].alter(10, 1500);
		atka.ent[1].alter(9, 1500);
		atka.ent[1].alter(10, 1500);

		if (atkTime == 0)
			anim.draw(gra, ori, siz);
		else {
			atka.draw(gra, ori, siz);
		}
	}

	@Override
	public int getAbi() {
		return 0;
	}

	@Override
	public int getDire() {
		return -1;
	}

	@Override
	public double getPos() {
		return b.ubase.pos + SNIPER_POS;
	}

	private void getAngle() {
		if (pos == -1) {
			if(preTime == 0 && bulletX == 0) {
				bulletAngle = 0;
			}

			targetAngle = 0;

			return;
		}

		int Cx = b.st.len * 4 - 3200;
		int Cy = 4400;

		int Ux = (int) (pos * 4);
		int Uy = 4480;

		double[] coords = bf.sniperCoords(preTime > 0);
		int scrollPos = (int) (-Math.round(coords[0] / CommonStatic.BattleConst.ratio * 4 / coords[1]));

		int sniperX = (Cx - scrollPos) / 10 + 203;
		int sniperY = (int) (Math.sin(Math.PI / 30 * b.time) * 10) + Cy / 10 - 369;

		int Uy10 = Uy / 10;
		int UyScroll = (Ux - scrollPos) / 10;
		double theta = Math.toDegrees(Math.atan2(sniperY - Uy10 - 4 * layer + 58, sniperX - UyScroll));

		if(preTime == 0 && bulletX == 0) {
			bulletAngle = theta;
		}

		targetAngle = theta;
	}

	public void update() {
		if (canDo && b.ubase.health <= 0) {
			canDo = false;
		}

		if (enabled && coolTime > 0)
			coolTime--;

		if (coolTime == 0 && enabled && pos > 0 && canDo) {
			if(Math.abs(targetAngle - cannonAngle) < 1) {
				coolTime = SNIPER_CD;
				preTime = SNIPER_PRE;
				atkTime = atka.len();
				atka.setup();
				anim.setup();
			} else {
				coolTime++;
			}
		}

		// find enemy pos
		if(preTime == 0) {
			pos = -1;

			for (Entity e : b.le)
				if (e.dire == 1 && e.pos > pos && !e.isBase && (e.touchable() & TCH_N) > 0) {
					target = e;
					pos = e.pos;
					layer = e.layer;
				}
		}

		if (enabled) {
			getAngle();
		}

		if (preTime > 0) {
			preTime--;
			if (preTime == 0) {
				//fire bullet
				bulletX = b.ubase.pos + SNIPER_POS + 375 * Math.cos(Math.toRadians((int) bulletAngle));

				atka.ent[6].alter(12, 1000);
				anim.ent[6].alter(12, 1000);
			}
		}

		if (bulletX != 0 && bulletX > pos) {
			bulletX = (int) (bulletX * 4 - 1500 * Math.cos(Math.toRadians((int) bulletAngle))) / 4.0;

			atka.ent[6].alter(4, (int) ((bulletX - b.ubase.pos - SNIPER_POS) / Math.cos(Math.toRadians((int) bulletAngle)) * CommonStatic.BattleConst.ratio * 0.75));
			anim.ent[6].alter(4, (int) ((bulletX - b.ubase.pos - SNIPER_POS) / Math.cos(Math.toRadians((int) bulletAngle)) * CommonStatic.BattleConst.ratio * 0.75));

			if (bulletX <= pos) {
				int atk = b.b.t().getBaseHealth() / 20;
				Proc proc = Proc.blank();
				proc.SNIPER.prob = 1;
				ArrayList<Trait> CTrait = new ArrayList<>();
				CTrait.add(UserProfile.getBCData().traits.get(TRAIT_TOT));
				AttackAb a = new AttackSimple(null, this, atk, CTrait, 0, proc, 0, getPos(), false, null, -1, true, 1);
				a.canon = -1;

				if(target != null && (target.touchable() & TCH_N) > 0) {
					target.damaged(a);
				}

				bulletX = 0;

				atka.ent[6].alter(12, 0);
				anim.ent[6].alter(12, 0);
			}
		}

		if (atkTime > 0) {
			atkTime--;
			atka.update(false);
		} else {
			anim.update(true);
		}

		if(bulletX > 0) {
			cannonAngle = targetAngle;
		} else {
			if(cannonAngle != targetAngle) {
				if(cannonAngle < targetAngle) {
					cannonAngle += 1;

					if(cannonAngle > targetAngle) {
						cannonAngle = targetAngle;
					}
				} else {
					cannonAngle -= 1;

					if(cannonAngle < targetAngle) {
						cannonAngle = targetAngle;
					}
				}
			}
		}

		if(bulletX > 0) {
			anim.ent[6].alter(12, 1000);
			anim.ent[5].alter(11, (int) Math.round(bulletAngle * 10));
		} else {
			anim.ent[5].alter(11, (int) Math.round(cannonAngle * 10));
		}

		atka.ent[5].alter(11, (int) Math.round(bulletAngle * 10));

		anim.ent[1].alter(5, - (int) Math.round((989.5 / 1.5 - 25 * Math.sin(Math.PI * b.time / 30) - height) * CommonStatic.BattleConst.ratio));
		atka.ent[1].alter(5, - (int) Math.round((989.5 / 1.5 - 25 * Math.sin(Math.PI * b.time / 30) - height) * CommonStatic.BattleConst.ratio));

		atka.ent[6].alter(9, 900);
		anim.ent[6].alter(9, 900);
		atka.ent[6].alter(10, 900);
		anim.ent[6].alter(10, 900);
	}

	public void cancel() {
		atkTime = 0;
		preTime = 0;
		bulletX = 0;
	}
}

File: battle/entity/Sniper.kt
package common.battle.entity

import common.battle.StageBasis
import common.battle.attack.AtkModelAb
import common.battle.attack.AttackAb
import common.battle.attack.AttackSimple
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.system.P
import common.system.fake.FakeGraphics
import common.util.Data
import common.util.Data.Proc
import common.util.anim.EAnimD
import common.util.pack.EffAnim.SniperEff
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter

class Sniper(sb: StageBasis?) : AtkModelAb(sb) {
    private val anim: EAnimD<*> = Data.Companion.effas().A_SNIPER.getEAnim(SniperEff.IDLE)
    private val atka: EAnimD<*> = Data.Companion.effas().A_SNIPER.getEAnim(SniperEff.ATK)
    private var coolTime: Int = Data.Companion.SNIPER_CD
    private var preTime = 0
    private var atkTime = 0
    private var angle = 0
    var enabled = true
    var canDo = true
    var pos = 0.0
    var height = 0.0
    var updown = 0.0

    /** attack part of animation  */
    fun drawAtk(g: FakeGraphics?, ori: P?, siz: Double) {
        // TODO
    }

    /** base part of animation  */
    fun drawBase(gra: FakeGraphics?, ori: P, siz: Double) {
        // TODO
        // double angle = Math.atan2(getPos() - pos, height);
        height = ori.y
        if (atkTime == 0) anim.draw(gra, ori, siz) else atka.draw(gra, ori, siz)
    }

    override fun getAbi(): Int {
        return 0
    }

    override fun getDire(): Int {
        return -1
    }

    override fun getPos(): Double {
        return b.st.len + Data.Companion.SNIPER_POS
    }

    fun update() {
        if (canDo && b.ubase.health <= 0) {
            canDo = false
        }
        if (enabled && coolTime > 0) coolTime--
        if (coolTime == 0 && enabled && pos > 0 && canDo) {
            coolTime = Data.Companion.SNIPER_CD
            preTime = Data.Companion.SNIPER_PRE
            atkTime = atka.len()
            atka.setup()
        }
        if (preTime > 0) {
            preTime--
            if (preTime == 0) {
                // attack
                val atk: Int = b.b.t().getBaseHealth() / 20
                val proc: Proc = Proc.Companion.blank()
                proc.SNIPER.prob = 1
                val a: AttackAb = AttackSimple(this, atk, -1, 0, proc, 0, getPos(), false, null)
                a.canon = -1
                b.getAttack(a)
            }
        }
        if (atkTime > 0) {
            atkTime--
            atka.update(false)
        } else {
            anim.update(true)
        }

        // find enemy pos
        pos = -1.0
        for (e in b.le) if (e.dire == 1 && e.pos > pos && !e.isBase && e.touchable() and Data.Companion.TCH_N > 0) pos = e.pos

        // Get angle of cannon and bullet
        angle = (-Math.toDegrees(Math.atan(height / (getPos() - pos)))).toInt() * 10
        anim.ent.get(5).alter(11, angle)
        atka.ent.get(5).alter(11, angle)

        // Get distance which bullet will fly
        // path = new P(-(getPos()-pos+300),height);
    }
}
